share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java

```diff
@@ -0,0 +1,523 @@
+
+            CoordinatorShardBuilderSupplier<ShareCoordinatorShard, CoordinatorRecord> supplier = () ->
+                new ShareCoordinatorShard.Builder(config);
+
+            CoordinatorEventProcessor processor = new MultiThreadedEventProcessor(
+                logContext,
+                "share-coordinator-event-processor-",
+                config.shareCoordinatorNumThreads(),
+                time,
+                coordinatorRuntimeMetrics
+            );
+
+            CoordinatorRuntime<ShareCoordinatorShard, CoordinatorRecord> runtime =
+                new CoordinatorRuntime.Builder<ShareCoordinatorShard, CoordinatorRecord>()
+                    .withTime(time)
+                    .withTimer(timer)
+                    .withLogPrefix(logPrefix)
+                    .withLogContext(logContext)
+                    .withEventProcessor(processor)
+                    .withPartitionWriter(writer)
+                    .withLoader(loader)
+                    .withCoordinatorShardBuilderSupplier(supplier)
+                    .withTime(time)
+                    .withDefaultWriteTimeOut(Duration.ofMillis(config.shareCoordinatorWriteTimeoutMs()))
+                    .withCoordinatorRuntimeMetrics(coordinatorRuntimeMetrics)
+                    .withCoordinatorMetrics(coordinatorMetrics)
+                    .withSerializer(new ShareCoordinatorRecordSerde())
+                    .withCompression(Compression.of(config.shareCoordinatorStateTopicCompressionType()).build())
+                    .withAppendLingerMs(config.shareCoordinatorAppendLingerMs())
+                    .build();
+
+            return new ShareCoordinatorService(
+                logContext,
+                config,
+                runtime,
+                coordinatorMetrics,
+                time
+            );
+        }
+    }
+
+    public ShareCoordinatorService(
+        LogContext logContext,
+        ShareCoordinatorConfig config,
+        CoordinatorRuntime<ShareCoordinatorShard, CoordinatorRecord> runtime,
+        ShareCoordinatorMetrics shareCoordinatorMetrics,
+        Time time) {
+        this.log = logContext.logger(ShareCoordinatorService.class);
+        this.config = config;
+        this.runtime = runtime;
+        this.shareCoordinatorMetrics = shareCoordinatorMetrics;
+        this.time = time;
+    }
+
+    @Override
+    public int partitionFor(String key) {
+        return Utils.abs(key.hashCode()) % numPartitions;
+    }
+
+    @Override
+    public Properties shareGroupStateTopicConfigs() {
+        Properties properties = new Properties();
+        properties.put(TopicConfig.CLEANUP_POLICY_CONFIG, TopicConfig.CLEANUP_POLICY_DELETE); // as defined in KIP-932
+        properties.put(TopicConfig.COMPRESSION_TYPE_CONFIG, BrokerCompressionType.PRODUCER.name);
+        properties.put(TopicConfig.SEGMENT_BYTES_CONFIG, config.shareCoordinatorStateTopicSegmentBytes());
+        return properties;
+    }
+
+    @Override
+    public void startup(
```

Comment:
```
The share coordinator startup method will get initialized from BrokerMetadataPublisher.
At the time of writing, the publisher uses metadata cache to fetch the number of partitions
of the share state topic. In case this information is not available, the user provided
config will be used to fetch the value. 
This is consistent with the group coordinator startup functionality.
```